### PR TITLE
Phx 2176 publish automation

### PR DESCRIPTION
**Pull request by cdwilhelm**
### Description

* use Raketask to create an active plugin list to be used for docs (later) and plugin policy
* upload the active plugin list to s3 (similar to policies)
* publish plugins merged to the master branch in the plugin publishing org.  

### Issues Resolved

[List any existing issues this PR resolves]

### Contribution Check List

- [ ] All tests pass.
- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable
- [ ] New functionality has been documented in CHANGELOG.MD

### CODE REVIEW

**Comment by cdwilhelm** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          2 Errors
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">tools/plugin_sync/README.md:1: MD002 First header should be a top level header<br />
tools/plugin_sync/README.md:12: MD029 Ordered list item prefix<br />
tools/plugin_sync/README.md:7: MD033 Inline HTML</p>

<p>A detailed description of the rules is available at https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">File syntax check failed. azure_aks_plugin.rb.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: tools/plugin_sync/plugin_sync.pt [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+      $file_name = last(split($item["file_name"],'/'))
+
+      # add the new plugin
+      $create_response = http_request(
+        auth: $$auth_rs,
+        verb: "post",
+        https: true,
+        host: 'cloud-3.rightscale.com',
+        href: join(["/cloud/v1/orgs/",$rs_org_id,"/plugins"]),
+        headers: { "api-version": "1.0" },
```
Comment (NeenuAVarghese) [inline tools/plugin_sync/plugin_sync.pt:163]: We do not need `api-version` header for the plugin api

File: tools/plugin_sync/plugin_sync.pt [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+      )
+      $response_array << to_s($create_response)
+      # set default-version
+      $version_response = http_request(
+        auth: $$auth_rs,
+        verb: "put",
+        https: true,
+        host: 'cloud-3.rightscale.com',
+        href: join(["/cloud/v1/orgs/",$rs_org_id,"/plugins/",$item['name'],"/default-version"]),
+        headers: { "api-version": "1.0" },
```
Comment (NeenuAVarghese) [inline tools/plugin_sync/plugin_sync.pt:174]: We do not need api-version header for the plugin api

File: tools/plugin_sync/plugin_sync.pt [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    }
+
+    if (force_upgrade == 1 ) {
+      var version_upgradeable = true
+    }
+
+    plugin_list.push({
+      name: active_plugin.name,
+      file_name: active_plugin.file_name,
+      repo_version: active_plugin.version,
```
Comment (NeenuAVarghese) [inline tools/plugin_sync/plugin_sync.pt:123]: should we name this plugin_version ?

File: tools/plugin_sync/plugin_sync.pt [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+  }
+
+  // loop through active plugins from s3 json file
+  for ( var i = 0; i < active_plugin_list.length; i++) {
+    var active_plugin = active_plugin_list[i]
+    // map catalog plugin with json list
+    var account_plugin = old_plugin_list[active_plugin.name]
+
+    var plugin_name = ""
+    var account_version = "0.0.0"
```
Comment (NeenuAVarghese) [inline tools/plugin_sync/plugin_sync.pt:90]: I am guessing that account_version is the version of plugin in the account ? Am I correct ?

File: Rakefile [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+        # skip plugin if the version isn't supplied or if version is '0.0'
+        if  ! publish
+          puts "Skipping #{name}, plugin not published"
+          next
+        end
+
+        puts "Adding #{name} #{file} #{version}"
+
+        file_list<<{
```
[the remark this answers is not in the document]
Comment (cdwilhelm) [reply Rakefile:60]: The dangerfile handles verifies those fields are set when the PR is created and updated.  They are also part of the manual review.

File: .travis.yml [hunk 6]
```diff
@@ -8,6 +8,18 @@ bundler_args: "--without documentation --path bundle"
...
 - git fetch --depth=1
 - gem install bundler
-- mkdir -p tmp && wget --no-check-certificate https://binaries.rightscale.com/rsbin/rsc/v8/rsc-linux-amd64.tgz && tar zxvf rsc-linux-amd64.tgz -C tmp;mv tmp/rsc/rsc . && chmod +x ./rsc && ./rsc --version 
+- mkdir -p tmp && wget --no-check-certificate https://binaries.rightscale.com/rsbin/rsc/v8/rsc-linux-amd64.tgz && tar zxvf rsc-linux-amd64.tgz -C tmp;mv tmp/rsc/rsc . && chmod +x ./rsc && ./rsc --version
 script:
 - bundle exec danger --verbose
+- bundle exec rake generate_plugin_list
+deploy:
+- provider: s3
+  bucket: rs-policysync-tool
```
[the remark this answers is not in the document]
Comment (cdwilhelm) [reply .travis.yml:17]: I thought of that, but the creds I have are scoped to the one bucket.  I thought it was easier to keep all the publishing automation together.

File: tools/plugin_sync/plugin_sync.pt (hunk 3 above)
Comment (cdwilhelm) [reply tools/plugin_sync/plugin_sync.pt:123]: There are two plugin version, one is the repo and one in the catalog.

File: tools/plugin_sync/plugin_sync.pt (hunk 4 above)
Comment (cdwilhelm) [reply tools/plugin_sync/plugin_sync.pt:90]: yes, that is the version that is in the catalog.

File: .travis.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (NeenuAVarghese) [reply .travis.yml:17]: Is it possible to separate policies vs plugins in different folders in the S3 bucket ?

File: tools/plugin_sync/plugin_sync.pt [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+      )
+      $plugin_source = $plugin_request["body"]
+      $file_name = last(split($item["file_name"],'/'))
+
+      # add the new plugin
+      $create_response = http_request(
+        auth: $$auth_rs,
+        verb: "post",
+        https: true,
+        host: 'cloud-3.rightscale.com',
```
[the remark this answers is not in the document]
Comment (cdwilhelm) [reply tools/plugin_sync/plugin_sync.pt]: Yes, will update

File: tools/plugin_sync/plugin_sync.pt (hunk 2 above)
Comment (cdwilhelm) [reply tools/plugin_sync/plugin_sync.pt]: will remove

File: tools/plugin_sync/plugin_sync.pt (hunk 1 above)
Comment (cdwilhelm) [reply tools/plugin_sync/plugin_sync.pt]: will remove

**Comment by cdwilhelm** [comment]
> Do we have automation around uploading these plugins to staging as well ?
@NeenuAVarghese   we don't, but we could set it up.  I don't think we have an org for the catalog in staging.  all we need to do is apply the policy on that org.

**Comment by NeenuAVarghese** [comment]
> @NeenuAVarghese we don't, but we could set it up. I don't think we have an org for the catalog in staging. all we need to do is apply the policy on that org.

We do have an org on staging !  https://github.com/rightscale/cloud_explorer/blob/master/_infra/staging/us-east-1/services/plugin_catalog_service.yaml#L145

File: .travis.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (cdwilhelm) [reply .travis.yml:17]: what is your concern about having them together?

[the remark this answers is not in the document]
Comment (NeenuAVarghese) [reply .travis.yml:17]: Sorry, I thought that we are uploading the plugins. Looks like we are just uploading the json. We can keep it in the same folder. Doesnt matter

[the remark this answers is not in the document]
Comment (cdwilhelm) [reply .travis.yml:17]: yep, it's just a single file of the latest plugin list which is used by the policy to compare what is in the catalog.  We also use the file to create an index in docs.

**Pull request merged**